app/app/explore/components/AddLiquidityModal.tsx

```diff
@@ -0,0 +1,274 @@
+import { useCallback, useEffect, useMemo, useState } from "react";
+import Image from "next/image";
+import Button from "@/app/button";
+import Modal from "@/app/modal";
+import { Balance, getBalance, PoolDetail as IPoolDetail } from "@/midgard";
+import { Slider } from "@shared/components/ui";
+import { twMerge } from "tailwind-merge";
+import {
+  getAssetShortSymbol,
+  getLogoPath,
+  isERC20,
+  normalizeAddress,
+} from "@/app/utils";
+import { useAppState } from "@/utils/context";
+import { useLiquidityPosition } from "@/hooks/useLiquidityPosition";
+import ErrorCard from "@/app/errorCard";
+import { useContracts } from "@/hooks/useContracts";
+
+interface AddLiquidityModalProps {
+  pool: IPoolDetail;
+  runePriceUSD: number;
+  onClose: () => void;
+}
+
+export default function AddLiquidityModal({
+  pool,
+  runePriceUSD,
+  onClose,
+}: AddLiquidityModalProps) {
+  const { wallet } = useAppState();
+  const { loading, error, addLiquidity } = useLiquidityPosition({ pool });
+  const [selectedTab] = useState("single");
+  const [assetAmount, setAssetAmount] = useState(0);
+  const [runeAmount, setRuneAmount] = useState(0);
+  const [txHash, setTxHash] = useState<string | null>(null);
+
+  // Initialize contract hooks
+  const poolViemAddress = pool.asset.split(".")[1].split("-")[1];
+  const tokenAddress = isERC20(pool.asset)
+    ? normalizeAddress(poolViemAddress)
+    : ""; // Set to empty string for non-ERC20 assets
+
+  // Use useContracts for ERC20 interaction
+  const {
+    balance: tokenBalance,
+    error: tokenError,
+    loadMetadata,
+  } = useContracts({
+    tokenAddress,
+    provider: wallet?.provider,
+  });
+
+  // Load token metadata on mount
+  useEffect(() => {
+    if (wallet?.provider) {
+      loadMetadata();
+    }
+  }, [wallet?.provider, loadMetadata]);
+
+  const getPercentage = (amount: number, max: number) => {
+    return max > 0 ? (amount / max) * 100 : 0;
+  };
+
+  const [runeBalance, setRuneBalance] = useState(0);
```

Comment:
imo this + getRuneBalance + the useEffect to refetch the balance every N secs should be in a custom hook. I don't believe there's a scenario where we would only want one of these things.